Lab4: - Коновалов Сергей 6204-010302D

seregaqt opened the pull request.

seregaqt committed:
Delete Readme.md

seregaqt committed:
Delete assignment.md

seregaqt committed:
Add files via upload

DiBoS290699 reviewed (commented):
# Результаты проверки текста отчета
## Рецензия на отчёт по лабораторной работе №4 (ООП на Java)

**1. Краткое резюме:**

Отчёт демонстрирует успешное выполнение большинства заданий курса, включая проектирование и реализацию классов для работы с функциями, их комбинациями и табулированием. Ключевым достоинством является следование структуре задания и попытка реализации всех запрошенных классов. Главный риск/пробел – отсутствие исходного кода, модульных тестов и подробного описания воспроизводимости результатов, что затрудняет проверку корректности и полноты реализации.

**2. Соответствие заданию:**

*   **Итог:** **Да**.
*   Отчёт описывает добавление конструкторов в `ArrayTabulatedFunction` и `LinkedListTabulatedFunction` (Задание 1).
*   Упоминается рефакторинг интерфейсов `TabulatedFunction` и создание `Function` (Задание 2).
*   Описано создание классов в пакетах `functions.basic` и `functions.meta` (Задания 3 и 4).
*   Сообщается о создании класса `Functions` со статическими методами (Задание 5).
*   Упоминается реализация методов табулирования в классе `TabulatedFunctions` (Задание 6).
*   Сообщается о добавлении методов ввода/вывода в `TabulatedFunctions` (Задание 7).
*   Упоминается реализация сериализации (Задание 9).

**3. Техническая корректность (ООП и Java):**

*   **Инкапсуляция:** Описано добавление конструкторов с проверкой входных данных, что говорит о попытке обеспечения инкапсуляции (по описанию).
*   **Интерфейсы:** Реализация интерфейса `Function` и расширение его интерфейсом `TabulatedFunction` соответствует принципам ООП (по описанию).
*   **Наследование:** Описано использование наследования для тригонометрических функций, что является хорошей практикой для избежания дублирования кода (по описанию).
*   **Композиция:** Создание классов `Sum`, `Mult` и других в пакете `functions.meta` предполагает использование композиции, что соответствует принципам ООП (по описанию).
*   **SOLID (SRP):** Структура пакетов (`functions.basic`, `functions.meta`) и разделение логики по классам указывает на следование принципу единственной ответственности (по описанию).
*   **Исключения:** Упоминается выбрасывание `IllegalArgumentException` при некорректных входных данных, что правильно с точки зрения обработки ошибок (по описанию).

**4. Архитектура и дизайн:**

*   **Модульная структура:** Разделение на пакеты `functions`, `functions.basic` и `functions.meta` демонстрирует модульную структуру.
*   **Паттерны:** Не указано, но можно предположить использование паттерна "Фабричный метод" в классе `Functions` (по описанию).
*   **Обоснование решений:** Отсутствует подробное обоснование выбора паттернов, интерфейсов и иерархий.

**5. Воспроизводимость (опционально):**

*   **Не представлено — допустимо.** Отсутствует информация об инструментах, версиях, командах сборки, структуре проекта и примерах запуска.

**6. Тестирование и результаты (опционально):**

*   **Не представлено — допустимо.** Отсутствует информация о модульных тестах, граничных случаях, негативных сценариях и интерпретации результатов.

**7. Стиль и конвенции (применимо при наличии кода):**

*   **Не применимо.** Код не предоставлен.

**8. Источники и добросовестность:**

*   **Не указано.** Отсутствуют ссылки на внешние материалы и источники.
*   **Подозрение.** Отсутствие кода и тестов не позволяет оценить оригинальность решения.

**9. Рекомендации по улучшению:**

*   **Добавить исходный код:** Предоставьте полный исходный код всех классов с комментариями.
*   **Написать модульные тесты:** Создайте тесты для проверки корректности работы конструкторов, методов и классов.
*   **Обосновать архитектурные решения:** Добавьте описание причин выбора конкретных паттернов, интерфейсов и иерархий.
*   **Описать воспроизводимость:** Предоставьте информацию об инструментах, версиях, командах сборки и примерах запуска.
*   **Улучшить ясность изложения:** Добавьте примеры использования классов и методов, чтобы сделать отчёт более понятным.
*   **Добавить UML-диаграммы:** Используйте UML-диаграммы для визуализации архитектуры и взаимосвязей классов.
*   **Добавить информацию об обработке исключений:** Опишите, как обрабатываются исключения в разных методах.


# Результаты проверки текста отчета на плагиаризм
Существенных совпадений не обнаружено.

# Результаты проверки кода

DiBoS290699 commented:
1. Необходимо использовать `StreamTokenizer`, а не `StringTokenizer`.
2. Больше комментариев в отчёте, не сплошной код.
3. В обоих случая сериализации в main вы используете `Externalizable` для `ArrayTabulatedFunction`, поэтому размер файлов совпал (что неверно для двух разных реализаций сериализации). Если класс реализует интерфейсы `Serializable` и `Externalizable` одновременно, то при сериализации выбирается `Externalizable` из-за высокой приоритетности.

seregaqt committed:
Исправил report, Main и TabulatedFunctions

DiBoS290699 reviewed (commented):
# Результаты проверки текста отчета
## Рецензия на отчёт по лабораторной работе №4 (ООП на Java)

**1. Краткое резюме**

Отчёт демонстрирует хорошее понимание принципов ООП и успешное выполнение заданий. Ключевым плюсом является реализованная и протестированная иерархия классов, охватывающая базовые функции, операции над ними и табулирование. Главный риск/пробел — недостаток пояснений по архитектурным решениям, особенно выбору интерфейсов и композиции вместо наследования в некоторых случаях.

**2. Соответствие заданию**

*   **Итог: Да**.
*   "Добавил конструкторы, принимающих массив точек, в оба класса табулированных функций." – Подтверждает выполнение Задания 1.
*   "Сделал рефакторинг иерархии классов, создав интерфейсы Function и TabulatedFunction." – Подтверждает выполнение Задания 2.
*   "Сделал реализацию основных математических функций в пакете functions.basic." – Подтверждает выполнение Задания 3.
*   "Сделал создание операций над функциями в пакете functions.meta." – Подтверждает выполнение Задания 4.

**3. Техническая корректность (ООП и Java)**

*   **Инкапсуляция:**  "Написал создание копий точек для обеспечения инкапсуляции данных." – Хорошая практика.
*   **Абстракция:** Интерфейс `Function` – эффективное использование абстракции для обобщения различных типов функций.
*   **Наследование:** Применение наследования для `Sin`, `Cos`, `Tan` от `TrigonometricFunction` оправдано.
*   **Полиморфизм:** Использование интерфейса `Function` позволяет гибко работать с различными реализациями функций.
*   **SOLID (SRP):** Классы пакетов `functions.basic` и `functions.meta` демонстрируют разделение ответственности.
*   **Исключения:** "Написал проверки на минимальное количество точек и упорядоченность по координате X." –  Правильное использование `IllegalArgumentException`. "Добавил обработку исключительных ситуаций." – Необходимо уточнить, какие именно ситуации обрабатываются.
*   **equals/hashCode/toString/compareTo:** Отсутствует — допустимо (не требовалось в задании).
*   **Управление ресурсами:**  "Добавил обработку исключений ввода-вывода." – Необходимо уточнить, как именно обрабатываются исключения (try-catch блоки, try-with-resources).
*   **Потокобезопасность:** Не релевантно.

**4. Архитектура и дизайн**

*   **Модульная структура:** Пакетная структура (`functions`, `functions.basic`, `functions.meta`) – хороший подход.
*   **Паттерны:** Отсутствуют — допустимо.
*   **UML-диаграммы:** Отсутствуют — допустимо.
*   **Обоснование:** Недостаточно обоснованы решения по использованию интерфейсов и композиции вместо наследования в `functions.meta`. Почему выбран именно такой подход для реализации операций над функциями?

**5. Воспроизводимость (опционально)**

*   Не представлено — допустимо.

**6. Тестирование и результаты (опционально)**

*   "Сделал комплексное тестирование всех написанных классов." — Хорошо, но не хватает деталей о тестовых сценариях (граничные случаи, негативные тесты).
*   "Провел сравнение точности и эффективности различных форматов хранения данных." —  Необходимо предоставить результаты сравнения (таблицы, графики).

**7. Стиль и конвенции (применимо при наличии кода)**

*   Код не предоставлен — допустимо.

**8. Источники и добросовестность**

*   Отсутствуют ссылки на внешние источники — допустимо.
*   Подозрения на заимствования отсутствуют.

**9. Рекомендации по улучшению**

*   **Добавьте UML-диаграммы:** Визуализация структуры классов поможет лучше понять архитектуру проекта.
*   **Обоснуйте архитектурные решения:** Объясните выбор интерфейсов и композиции вместо наследования в `functions.meta`. Почему было выбрано именно такое решение?
*   **Предоставьте тестовые примеры:** Добавьте несколько примеров тестовых сценариев, особенно для граничных случаев и негативных тестов.
*   **Детализируйте обработку исключений:**  Опишите, как обрабатываются исключения ввода-вывода и какие ситуации обрабатываются в конструкторах.
*   **Улучшите ясность изложения:**  Добавьте больше пояснений к коду и логике работы классов.
*   **Рассмотрите возможность использования логгера:** Вместо вывода в консоль, можно использовать логгер для более структурированного вывода информации о работе программы.
*   **Если будете добавлять модульные тесты:** Убедитесь, что тесты покрывают основные сценарии и граничные случаи.


# Результаты проверки текста отчета на плагиаризм
Существенных совпадений не обнаружено.

# Результаты проверки кода
## Отчёт по лабораторной работе

### Работоспособность

Код в целом соответствует поставленной задаче и требованиям лабораторной работы. Все реализованные классы и методы выполняют свои функции согласно спецификации.  В частности:

*   Конструкторы `ArrayTabulatedFunction` и `LinkedListTabulatedFunction` корректно выбрасывают исключения `IllegalArgumentException` при некорректных входных данных (меньше двух точек, неупорядоченные точки).
*   Интерфейс `TabulatedFunction` расширен интерфейсом `Function`, методы `getLeftDomainBorder()`, `getRightDomainBorder()`, `getFunctionValue(double x)` удалены из `TabulatedFunction`.
*   Классы `Exp`, `Log`, `Sin`, `Cos`, `Tan` реализуют интерфейс `Function` и корректно вычисляют соответствующие значения.
*   Классы `Sum`, `Mult`, `Power`, `Scale`, `Shift`, `Composition` реализуют интерфейс `Function` и комбинируют функции, как требуется.
*   Класс `Functions` предоставляет статические методы для комбинирования функций, используя классы из пакета `functions.meta`.
*   Класс `TabulatedFunctions` содержит статические методы для табулирования и сериализации/десериализации функций.

Однако, в классе `LinkedListTabulatedFunction` при добавлении точки, не проверяется, что новая точка не дублирует существующие. Это может привести к проблемам при дальнейшей работе с функцией.

### Качество кода

В целом качество кода хорошее, но есть некоторые моменты, которые можно улучшить:

*   **Читаемость и понятность:** Названия переменных в большинстве случаев понятны и информативны. Комментарии присутствуют, но могли бы быть более подробными в некоторых местах.
*   **Простота и избегание избыточности:** Код достаточно структурирован.  Однако, в классе `LinkedListTabulatedFunction` много повторяющегося кода, например, в методах поиска и обновления точек. Можно вынести общую логику в отдельные вспомогательные методы.
*   **Соответствие базовым рекомендациям по стилю:** Код в целом соответствует базовым рекомендациям по стилю Java.
*   **Инкапсуляция:** В конструкторах `ArrayTabulatedFunction` и `LinkedListTabulatedFunction` создаются копии входных данных, что обеспечивает инкапсуляцию.
*   **Обработка исключений:**  В методах ввода/вывода данных в `TabulatedFunctions` выбрасываются исключения `RuntimeException`, что может быть не самым лучшим подходом. Лучше использовать более специфические исключения, например `IOException`.
*   **Потоки:** Использование `DataInputStream` и `DataOutputStream` для сериализации в бинарном формате логично и эффективно. Использование `StreamTokenizer` для чтения из текстового потока также правильно.

**Пример проблемного участка кода:**

В классе `LinkedListTabulatedFunction`, метод `addPoint()` не проверяет, существует ли точка с таким же x-координатой, что может привести к дублированию точек.

```java
public void addPoint(FunctionPoint point) throws InappropriateFunctionPointException {
    FunctionNode currentNode = head.getNext();
    for (int i = 0; i < pointsCount; i++) {
        if (doubleEquals(point.getX(), currentNode.getPoint().getX())) {
            throw new InappropriateFunctionPointException

seregaqt committed:
Update Main.java

Изменил Main, без добавления нового класса, ArrayTabulatedFunction - Externalizable, LinkedList - Serializable

DiBoS290699 reviewed (commented):
# Результаты проверки текста отчета
## Рецензия на отчёт по лабораторной работе №4 (ООП на Java)

**1. Краткое резюме:**

Отчёт демонстрирует хорошее понимание принципов ООП и успешную реализацию поставленной задачи по созданию системы табулированных функций. Ключевым плюсом является полное выполнение всех заданий и подробное описание реализованных классов и методов. Главный риск/пробел – отсутствие исходного кода и тестов, что затрудняет проверку корректности и качества реализации.

**2. Соответствие заданию:**

*   **Итог: Да**.
*   В отчете подробно описаны конструкторы классов `ArrayTabulatedFunction` и `LinkedListTabulatedFunction` с проверками на корректность входных данных (минимальное количество точек и упорядоченность).
*   Интерфейсы `Function` и `TabulatedFunction` реализованы в соответствии с требованиями задания.
*   Классы `Exp`, `Log`, `Sin`, `Cos`, `Tan` корректно реализуют интерфейс `Function` и учитывают особенности области определения каждой функции.
*   Классы `Sum`, `Mult`, `Power`, `Scale`, `Shift`, `Composition` реализованы в пакете `functions.meta` и позволяют комбинировать функции.

**3. Техническая корректность (ООП и Java):**

*   **Инкапсуляция:** Упоминается создание копий точек в конструкторах `ArrayTabulatedFunction` и `LinkedListTabulatedFunction` для обеспечения инкапсуляции данных (по описанию).
*   **Абстракция:** Интерфейс `Function` абстрагирует понятие функции, позволяя реализовать различные типы функций.
*   **Наследование:** Классы `Sin`, `Cos`, `Tan` наследуют от базового класса `TrigonometricFunction`, что демонстрирует использование наследования для избежания дублирования кода.
*   **Исключения:** Реализована обработка исключительных ситуаций при создании конструкторов и табулировании функций (по описанию).
*   **Классы/интерфейсы:**  Правильное использование интерфейсов и классов для организации структуры проекта.
*   **Композиция vs наследование:** Использование композиции в классе `Composition` для объединения функций.

**4. Архитектура и дизайн:**

*   **Модульная структура:**  Проект имеет четкую модульную структуру с пакетами `functions`, `functions.basic` и `functions.meta`.
*   **Паттерны:**  Использование паттерна "фабрика" в классе `Functions` для создания объектов функций.
*   **Обоснование решений:**  Указано обоснование использования наследования для тригонометрических функций и композиции для объединения функций.

**5. Воспроизводимость (опционально):**

*   Не представлено — допустимо.

**6. Тестирование и результаты (опционально):**

*   Не представлено — допустимо.

**7. Стиль и конвенции (применимо при наличии кода):**

*   Код не предоставлен — допустимо.

**8. Источники и добросовестность:**

*   Не указаны ссылки на внешние материалы/шаблоны/библиотеки.

**9. Рекомендации по улучшению:**

*   **Добавьте исходный код:**  Предоставление исходного кода позволит проверить корректность реализации и оценить качество написанного кода.
*   **Напишите модульные тесты:**  Модульные тесты помогут убедиться в правильной работе каждого класса и метода.
*   **Дополните описание архитектуры:**  Более подробное описание архитектуры проекта (например, UML-диаграммы) поможет лучше понять структуру и взаимосвязи между классами.
*   **Укажите версии используемых инструментов:**  Добавьте информацию о версии JDK и используемых библиотек.
*   **Добавьте примеры запуска:**  Предоставьте примеры запуска программы и ожидаемый вывод для демонстрации работы реализованных классов и методов.
*   **Улучшите ясность изложения:**  Сделайте акцент на обосновании ключевых архитектурных решений (например, почему выбран паттерн "фабрика", а не другой).


# Результаты проверки текста отчета на плагиаризм
Существенных совпадений не обнаружено.

# Результаты проверки кода

DiBoS290699 commented:
У вас опять в обоих случаях тестируется Externalizable, только уже для разных классов. Напоминаю:

> Если класс реализует интерфейсы `Serializable` и `Externalizable` одновременно, то при сериализации выбирается `Externalizable` из-за высокой приоритетности.

P.s. не забудьте потом обновить скриншоты в отчёте

seregaqt committed:
Array, LinkedList, Main, report

Убрал ненужные сериализации из implements функций Array и LinkedList, отредактировал Main и изменил report

DiBoS290699 commented:
Принято
